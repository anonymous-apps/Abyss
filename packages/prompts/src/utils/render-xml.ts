import type { CellXMLElement } from '../promptTemplate.types';

export function renderXmlCell(cell: CellXMLElement): string {
    return `\n${handleObject(cell.content).join('\n')}\n`;
}

function renderInlineString(key: string, value: string, indent: number): string {
    return `${'  '.repeat(indent)}<${key}>${value}</${key}>`;
}

function renderMultilineString(key: string, value: string, indent: number): string {
    return `${'  '.repeat(indent)}<${key}>\n${'  '.repeat(indent + 1)}<!CDATA[${value}]>\n${'  '.repeat(indent)}</${key}>`;
}

function renderBoolean(key: string, value: boolean, indent: number): string {
    return `${'  '.repeat(indent)}<${key} boolean="${value ? 'true' : 'false'}" />`;
}

function renderNumber(key: string, value: number, indent: number): string {
    return `${'  '.repeat(indent)}<${key} number="${value}" />`;
}

function renderNull(key: string, indent: number): string {
    return `${'  '.repeat(indent)}<${key} null="true" />`;
}

function renderObject(key: string, value: object, indent: number): string[] {
    const lines: string[] = [];
    lines.push(`${'  '.repeat(indent)}<${key}>`);
    lines.push(...handleObject(value, indent + 1));
    lines.push(`${'  '.repeat(indent)}</${key}>`);
    return lines;
}

const handleObject = (obj: object, indent = 0): string[] => {
    const lines: string[] = [];
    const keys = Object.keys(obj);
    const object = obj as Record<string, unknown>;

    for (const key of keys) {
        const value = object[key];
        if (value === undefined) {
            continue;
        }
        if (value === null) {
            lines.push(renderNull(key, indent));
            continue;
        }
        if (typeof value === 'boolean') {
            lines.push(renderBoolean(key, value, indent));
            continue;
        }
        if (typeof value === 'number') {
            lines.push(renderNumber(key, value, indent));
            continue;
        }
        if (typeof value === 'string') {
            if (value.includes('\n')) {
                lines.push(renderMultilineString(key, value, indent));
            } else {
                lines.push(renderInlineString(key, value, indent));
            }
            continue;
        }
        if (typeof value === 'object') {
            lines.push(...renderObject(key, value, indent));
            continue;
        }

        throw new Error(`Unsupported value type: ${typeof value}`);
    }
    return lines;
};
