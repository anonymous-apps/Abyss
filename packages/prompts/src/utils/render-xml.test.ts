import { describe, expect, it } from 'vitest';
import { renderXmlCell } from './render-xml';

describe('renderXmlCell', () => {
    it('should render a simple xml element', () => {
        const result = renderXmlCell({
            type: 'xmlElement',
            content: {
                name: 'test',
                value: 'test',
            },
        });
        expect(result).toMatchSnapshot();
    });

    it('should render a nested xml element', () => {
        const result = renderXmlCell({
            type: 'xmlElement',
            content: {
                name: 'test',
                value: {
                    childName: 'nestedTest',
                    childValue: 'nestedValue',
                },
            },
        });
        expect(result).toMatchSnapshot();
    });

    it('should render multiple sibling elements', () => {
        const result = renderXmlCell({
            type: 'xmlElement',
            content: {
                item1: 'first',
                item2: 'second',
                item3: 'third',
            },
        });
        expect(result).toMatchSnapshot();
    });

    it('should render deeply nested xml elements', () => {
        const result = renderXmlCell({
            type: 'xmlElement',
            content: {
                level1: {
                    level2: {
                        level3: {
                            text: 'deep value',
                        },
                    },
                },
            },
        });
        expect(result).toMatchSnapshot();
    });

    it('should handle string values with dedent', () => {
        const multilineString = 'This is a\nmultiline string.';
        const result = renderXmlCell({
            type: 'xmlElement',
            content: {
                data: multilineString,
            },
        });
        expect(result).toMatchSnapshot();
    });

    it('should render non-string primitive values', () => {
        const result = renderXmlCell({
            type: 'xmlElement',
            content: {
                count: 123,
                isActive: true,
                isNull: null,
                isUndefined: undefined,
            },
        });
        expect(result).toMatchSnapshot();
    });

    it('should render an empty content object as an empty string with newlines', () => {
        const result = renderXmlCell({
            type: 'xmlElement',
            content: {},
        });
        expect(result).toMatchSnapshot();
    });
});
