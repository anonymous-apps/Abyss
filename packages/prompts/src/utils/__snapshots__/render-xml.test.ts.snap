// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderXmlCell > should handle string values with dedent 1`] = `
"
<data>
  <!CDATA[This is a
multiline string.]>
</data>
"
`;

exports[`renderXmlCell > should render a nested xml element 1`] = `
"
<name>test</name>
<value>
  <childName>nestedTest</childName>
  <childValue>nestedValue</childValue>
</value>
"
`;

exports[`renderXmlCell > should render a simple xml element 1`] = `
"
<name>test</name>
<value>test</value>
"
`;

exports[`renderXmlCell > should render an empty content object as an empty string with newlines 1`] = `
"

"
`;

exports[`renderXmlCell > should render deeply nested xml elements 1`] = `
"
<level1>
  <level2>
    <level3>
      <text>deep value</text>
    </level3>
  </level2>
</level1>
"
`;

exports[`renderXmlCell > should render multiple sibling elements 1`] = `
"
<item1>first</item1>
<item2>second</item2>
<item3>third</item3>
"
`;

exports[`renderXmlCell > should render non-string primitive values 1`] = `
"
<count number="123" />
<isActive boolean="true" />
<isNull null="true" />
"
`;
